rfcs/56-better-reflect.md

```diff
@@ -0,0 +1,310 @@
+- `any`
+- `any_mut`
+- `downcast`
+- `take`
+- `is`
+- `downcast_ref`
+- `downcast_mut`
+- `set`
+
+The `Any` trait bound on `Reflect` is similarly problematic, as it introduces
+the same issues as those methods.
+
+Note that the following methods are also problematic, and require discussion,
+but to keep to scope of this RFC short, we will keep them in `Reflect`.
+- `reflect_hash`
+- `reflect_partial_eq`
+- `serializable`
+
+## Proposition
+
+- Create a `CanonReflect` trait
+- Remove all methods assuming uniqueness and `Any` trait bound from `Reflect`,
+  add them to `CanonReflect`.
+- Remove `set` method from `Reflect`
+- Implement a way to convert from `Reflect` to `CanonReflect`:
+  - New method to `Typed` trait: `from_reflect`
+  - Remove `FromReflect` trait
+  - New `constructor` field to `TypeRegistration` to convert a `Box<dyn Reflect>`
+    into a `Box<dyn CanonReflect>` dynamically.
+  - New `make_canon` method to `TypeRegistry`
+
+### `CanonReflect` trait
+
+We remove all methods assuming uniqueness and `Any` trait bound from
+`Reflect` and only implement them on "canonical" types: ie: the ones implemented
+by the `#[derive(Reflect)]` macros.
+
+We introduce a new trait `CanonReflect`, that must be explicitly implemented
+to mark a type as being the canonical type.
+```rust
+pub trait CanonReflect: Reflect + Any {
+  fn any_mut(&mut self) -> &mut dyn Any {
+    // implementation
+  }
+  // etc.
+}
+```
+
+This trait is automatically implemented by the `#[derive(Reflect)]` macro.
+
+This prevents any surprises caused by the underlying type of a reflect not being
+the one expected.
+
+Note that when [trait upcasting] is implemented, the `any_mut` and other
+`Any`-related methods should be moved into a `impl dyn CanonReflect` block.
+
+### `Reflect` to `CanonReflect` conversion
+
+We still want to be able to use the `Any` methods on our `Reflect`, so we want
+a way to get a `CanonReflect` from a `Reflect`. This is only possible by
+checking if the underlying type is the one we expect OR converting into the
+underlying type described by the `Reflect::type_name()` method.
+
+To do so, it's necessary to either provide an expected type or access the type
+registry to invoke the constructor associated with the `Reflect::type_name()`.
+
+To define a constructor We move the `from_reflect` method from `FromReflect` to
+`Typed`. For no particular reasons but to reduce how many traits we export from
+`bevy_reflect`.
+
+**Danger**: Using `type_name` from `Reflect` is error-prone. For the conversion
+to pick up the right type, we need `type_name()` to match exactly the one
+returned by `std::any::type_name`. Ideally we have an alternative that defines
+the underlying type without ambiguity, such as `target_type_id` or
+`full_type_name`.
+
+We introduce a new field to `TypeRegistration`: `constructor`. It allows
+converting from a `Reflect` to a `CanonReflect`, so that it's possible to
+use its underlying-sensible methods.
```

Comment:
Yeah. I think we can avoid using `FromReflect` for this, if we add a `into_canon` method to `PartialReflect`.

Reply:
It's still problematic though. Since there is no way to get a `CanonReflect` from a `Reflect` "proxy" this way. The only way I can see is if we add the `FromReflect::from_reflect` method to the type registry the same way we do for other traits.